File: tests/src/main/java/fr/pgreze/testme/ui/MainActivity.java
package fr.pgreze.testme.ui;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;

import fr.pgreze.testme.domain.common.di.HasComponent;
import fr.pgreze.testme.ui.common.di.ActivityComponent;
import fr.pgreze.testme.ui.common.di.DaggerActivityComponent;
import fr.pgreze.testme.ui.gist.PublicGistsFragment;

public class MainActivity extends AppCompatActivity implements HasComponent<ActivityComponent> {

    private ActivityComponent component;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState == null) {
            getSupportFragmentManager().beginTransaction()
                    .add(android.R.id.content, new PublicGistsFragment(), PublicGistsFragment.TAG)
                    .commit();
        }
    }

    @Override
    public ActivityComponent component() {
        if (component == null) {
            component = DaggerActivityComponent.builder()
                    .appComponent(TestApp.getInstance(this).component())
                    .build();
        }
        return component;
    }
}

File: dagger/android/src/main/kotlin/pgreze/dagger2/ui/MainActivity.kt
package pgreze.dagger2.ui

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import pgreze.dagger2.R
import pgreze.dagger2.app.Dagger2App
import pgreze.dagger2.di.AndroidLifecycleScope
import pgreze.dagger2.di.HasComponent

class MainActivity : AppCompatActivity(), HasComponent<ActivityComponent> {

    /**
     * This component is [AndroidLifecycleScope] annotated,
     * so it will be tight to the activity lifecycle.
     *
     * In other words, when activity will be destroyed,
     * all [AndroidLifecycleScope] objects will be destroyed as well (no memory leak).
     */
    override val component: ActivityComponent by lazy {
        DaggerActivityComponent.builder()
                .appComponent(Dagger2App.getInstance(this).component)
                .build()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
    }
}
